### ISSUE TITLE
`socket`/`tcp` source: `max_length` configuration discards messages instead of truncating

### ISSUE THREAD

**Issue by hhromic** [vectordotdev/vector#11512]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Problem

According to the [official documentation](https://vector.dev/docs/reference/configuration/sources/socket/#max_length) for the `max_length` configuration in the `socket` source:

> The maximum buffer size of incoming messages. Messages larger than this are truncated.

However, when the socket is in `tcp` mode, a larger message gets discarded, not truncated. In `udp` mode the behaviour is correct.

For example, consider the Vector configuration shown below, and start Vector:
```
$ vector -c vector.yaml
2022-02-22T15:51:38.899961Z  INFO vector::app: Log level is enabled. level="vector=info,codec=info,vrl=info,file_source=info,tower_limit=trace,rdkafka=info,buffers=info"
2022-02-22T15:51:38.900265Z  INFO vector::app: Loading configs. paths=["vector.yaml"]
2022-02-22T15:51:38.907676Z  INFO vector::topology::running: Running healthchecks.
2022-02-22T15:51:38.908138Z  INFO vector::topology::running: Starting source. key=socket
2022-02-22T15:51:38.908514Z  INFO vector::topology::running: Starting sink. key=sink0
2022-02-22T15:51:38.909878Z  INFO vector: Vector has started. debug="false" version="0.20.0" arch="x86_64" build_id="2a706a3 2022-02-11"
2022-02-22T15:51:38.910088Z  INFO vector::app: API is disabled, enable by setting `api.enabled` to `true` and use commands like `vector top`.
2022-02-22T15:51:38.914960Z  INFO vector::topology::builder: Healthcheck: Passed.
2022-02-22T15:51:38.925094Z  INFO source{component_kind="source" component_id=socket component_type=socket component_name=socket}: vector::sources::util::tcp: Listening. addr=0.0.0.0:8080
```

Then, in another terminal, send a simple TCP message larger than 10 bytes using `socat`:
```
$ echo "hello world" | socat -u -dd stdin tcp-connect:localhost:8080
2022/02/22 15:51:41 socat[4504] N using stdin for reading
2022/02/22 15:51:41 socat[4504] N opening connection to AF=2 127.0.0.1:8080
2022/02/22 15:51:41 socat[4504] N successfully connected from local address AF=2 127.0.0.1:36892
2022/02/22 15:51:41 socat[4504] N starting data transfer loop with FDs [0,0] and [5,5]
2022/02/22 15:51:41 socat[4504] N socket 1 (fd 0) is at EOF
2022/02/22 15:51:41 socat[4504] N exiting with status 0
```

Now observe how Vector discarded the message instead of truncating as stated in the documentation:
```
2022-02-22T15:51:41.809346Z  WARN source{component_kind="source" component_id=socket component_type=socket component_name=socket}:connection{peer_addr=127.0.0.1:36892}: vector::codecs::decoding::framing::character_delimited: Discarding frame larger than max_length. buf_len=12 max_length=10 internal_log_rate_secs=30
```

As mentioned, in `udp` mode, the message does get truncated as documented:
```
2022-02-22T15:47:53.006849Z  INFO source{component_kind="source" component_id=socket component_type=socket component_name=socket}: vector::sources::socket::udp: Listening. address=0.0.0.0:8080
{"host":"127.0.0.1:51050","message":"hello worl","source_type":"socket","timestamp":"2022-02-22T15:47:59.297358700Z"}
```

### Configuration

```text
sources:
  socket:
    type: socket
    mode: tcp
    address: 0.0.0.0:8080
    max_length: 10

sinks:
  sink0:
    type: console
    inputs:
      - socket
    encoding: json
```

### Version

vector 0.20.0 (x86_64-unknown-linux-gnu 2a706a3 2022-02-11)

**Comment by StephenWakely** [comment]
With the tcp socket source if the message exceeds max length the message is dropped. With udp the message get's truncated. 

The documentation says the message should be truncated. This isn't a regression, I can replicate this in v0.11.

I have three options:
1. Make it so tcp also truncates.
2. Make it so udp also drops the message - change the documentation.
3. Change the documentation to say tcp drops and udp truncates.

@hhromic do you have any thoughts/preferences?

**Comment by hhromic** [comment]
Hi @StephenWakely! Mmm interesting choices.
I (and hopefully most users) would prefer for Vector to be consistent, so Option (3) I think should be left out.
Options (1) or (2) makes Vector to become "opinionated" about _what to do_ instead of letting the user choose.

I don't think we want Vector to start creeping on configuration parameters, so if Vector is to be opinionated on this regard, and given that one of Vector's primary goals is reliability, I would think that truncation is preferred over dropping. I.e., part of the messages is better than losing the message.

So in summary, I lean on Option (1).

**Comment by StephenWakely** [comment]
> So in summary, I lean on Option (1).

Thanks for the feedback @hhromic. We have instead decided to go with option 2. Option 2 is already implemented in our Decoders so moving it to option 1 would have a bigger impact on existing functionality. It seems the UDP source doesn't use those decoders, hence the inconsistency.

I will raise an issue to see if we can make this a customisable option.

**Comment by hhromic** [comment]
> Option 2 is already implemented in our Decoders so moving it to option 1 would have a bigger impact on existing functionality.

That's a quite good argument, sounds good then! Dropped events can still be accounted in metrics so we can always monitor that and investigate if it increases. Cheers!
